impl/src/main/java/org/pentaho/di/trans/dataservice/streaming/execution/StreamingServiceTransExecutor.java

```diff
@@ -0,0 +1,241 @@
+   */
+  public StreamExecutionListener getBuffer( String query, int windowSize, long windowMillis ) {
+    String cacheId = getCacheKey( query, windowSize, windowMillis );
+
+    StreamExecutionListener streamListener = serviceListeners.getIfPresent( cacheId );
+
+    if ( streamListener == null ) {
+      if ( stepStream == null ) {
+        stepStream = new StreamList<>();
+      }
+
+      Observable<List<RowMetaAndData>> buffer = windowMillis > 0
+        ? windowSize > 0 ? stepStream.getStream().buffer( windowMillis, TimeUnit.MILLISECONDS, windowSize )
+        : stepStream.getStream().buffer( windowMillis, TimeUnit.MILLISECONDS )
+        : stepStream.getStream().buffer( windowSize );
+
+      streamListener = new StreamExecutionListener( buffer );
+
+      serviceListeners.put( cacheId, streamListener );
+    }
+
+    if ( !isRunning.get() ) {
+      isRunning.compareAndSet( false, true );
+      startService();
+    }
+
+    return streamListener;
+  }
+
+  /**
+   * Generates the cache key for a given query with a specific size and rate.
+   *
+   * @param query The query.
+   * @param size The query window size.
+   * @param millis The query window rate.
+   * @return The cache key for the query.
+   */
+  private String getCacheKey( String query, int size, long millis ) {
+    return query.concat( String.valueOf( size ) ).concat( String.valueOf( millis ) );
+  }
+
+  /**
+   * Starts the Service transformation and its row event listener.
+   */
+  private void startService() {
+    try {
+      lastCacheCleanupMillis = new Date().getTime();
+      serviceTrans.startThreads();
+
+      StepInterface serviceStep = serviceTrans.findRunThread( serviceStepName );
+      if ( serviceStep == null ) {
+        throw Throwables.propagate( new KettleException( "Service step is not accessible" ) );
+      }
+      serviceStep.addRowListener( new RowAdapter() {
+        /**
+         * Listener for the service transformation output rows.
+         * If the stepStream has any valid registered listeners it writes the output row to the stepStream,
+         * otherwise stops the service transformation and kills its running thread.
+         *
+         * @param rowMeta The metadata of the written row.
+         * @param row The data of the written row.
+         * @throws KettleStepException
+         */
+        @Override
+        public void rowWrittenEvent( RowMetaInterface rowMeta, Object[] row ) throws KettleStepException {
+          // Simply pass along the row to the other transformation (to the Injector step)
+          LogChannelInterface log = serviceTrans.getLogChannel();
+
+          try {
+            if ( log.isRowLevel() ) {
+              log.logRowlevel( DataServiceConstants.PASSING_ALONG_ROW + rowMeta.getString( row ) );
+            }
+          } catch ( KettleValueException e ) {
+            // Ignore error
+          }
+
+          RowMetaAndData rowData = new RowMetaAndData();
```

Comment:
Shouldn't the rowData field be inside the serviceListeners.size() > 0 if below, since it isn't used outside of that scope?